File: src/prompts.ts
/**
 * Prompt builders — ported from Python getting_started/examples/langchain/app.py.
 *
 * buildGreeting        → ConversationRelay welcome_greeting for outbound calls
 * buildSystemPrompt    → LLM system prompt for outbound calls
 * buildInboundSystemPrompt → LLM system prompt for inbound calls (uses profile traits)
 * buildMemoryContext   → formats TACMemoryResponse as a text block for AgentCore instruction
 */

import { MemberProfile, TACMemoryResponse } from './types';

/** Spoken greeting injected into TwiML for outbound calls. */
export function buildGreeting(name: string, goal: string, _goalDesc: string): string {
  const topic = goal || 'your care plan';
  return (
    `Hi ${name}, this is the Owl Health Care Team calling. ` +
    `I'm reaching out about ${topic}. ` +
    `Do you have a moment to chat?`
  );
}

/** System prompt for an outbound call — gives the agent purpose and member context. */
export function buildSystemPrompt(name: string, goal: string, goalDesc: string): string {
  const topic = goal || 'their care plan';
  const followUpDetail = goalDesc
    ? `Once they respond, use the following to guide your follow-up questions: ${goalDesc}`
    : '';
  return (
    `You are an Owl Health care coordination agent making an outbound call to ${name}. ` +
    `The member has already been greeted and told this call is about ${topic} — do not re-introduce yourself or repeat that. ` +
    `Wait for their response, then naturally guide the conversation using what you know. ` +
    `${followUpDetail} ` +
    `Be warm and conversational — speak in plain sentences, no bullet points, no bold text, no special formatting. ` +
    `Keep responses brief and easy to follow on a phone call. ` +
    `Guide the conversation toward a clear next step or action.`
  ).trim();
}

/** System prompt for an inbound call — personalised from profile traits. */
export function buildInboundSystemPrompt(profile: MemberProfile | null): string {
  let name: string | null = null;
  let nextFollowUp: string | null = null;
  let nextFollowUpReason: string | null = null;

  if (profile?.traits) {
    const contact = profile.traits.Contact;
    const outreach = profile.traits.outreach;
    if (contact?.firstName) {
      name = [contact.firstName, contact.lastName].filter(Boolean).join(' ');
    }
    nextFollowUp = outreach?.nextFollowUp ?? null;
    nextFollowUpReason = outreach?.nextFollowUpReason ?? null;
  }

  let prompt =
    'You are an Owl Health care coordination agent handling an inbound call. ' +
    'Be warm and conversational — speak in plain sentences, no bullet points, no bold text, no special formatting. ' +
    'Keep responses brief and easy to follow on a phone call. ' +
    'You have been provided with the member\'s profile and past call summaries — ' +
    'use this to personalize your responses and avoid asking for information you already have.';

  if (name) prompt += ` The member's name is ${name}.`;
  if (nextFollowUp) prompt += ` Their next scheduled follow-up is about: ${nextFollowUp}.`;
  if (nextFollowUpReason) prompt += ` Use this to guide your questions: ${nextFollowUpReason}.`;

  return prompt;
}

/**
 * Format TAC Conversation Memory (observations + summaries) as a text block.
 * This is injected into AgentCore's `instruction` field, mirroring Python's
 * MemoryPromptBuilder which prepends a SystemMessage with "# Customer Context".
 */
export function buildMemoryContext(memory: TACMemoryResponse | null): string | null {
  if (!memory) return null;

  const sections: string[] = [];

  if (memory.observations.length > 0) {
    sections.push(
      '## Key Observations\n' +
      'Important notes about the customer from previous interactions:\n' +
      memory.observations.map((o) => `- ${o}`).join('\n'),
    );
  }

  if (memory.summaries.length > 0) {
    sections.push(
      '## Previous Call Summaries\n' +
      memory.summaries.map((s) => `- ${s}`).join('\n'),
    );
  }

  if (sections.length === 0) return null;

  return (
    '# Customer Context\n' +
    'You have access to the following information about this customer from previous interactions:\n\n' +
    sections.join('\n\n')
  );
}

File: src/tac-hls-server/prompts.py
"""System prompt builders for each channel."""

import pathlib

_DIR = pathlib.Path(__file__).parent

BASE_SYSTEM_PROMPT = """You are an Owl Health care coordination agent. Your role is to handle member calls on behalf of the care team. Your tone is warm, empathetic, and conversational.

# OPERATIONAL GUIDELINES
- Communication Style: Speak in plain, natural sentences.
- Formatting Restrictions: DO NOT use bullet points, bold text, italics, or any special Markdown formatting. Each response must be brief and optimized for a voice/phone call.
- Context Awareness: Do not re-introduce yourself or ask for the member's time, as they have already been greeted. Use provided call context to personalize responses.
- Data Efficiency: Never ask for information that is already present in the member's profile or previous call summaries.

# HUMAN TRANSFER
- If the member asks to speak to a human, agent, representative, or care specialist, acknowledge their request warmly and naturally in your response, then call the escalate_to_human function.
- If you detect distress or a safety risk (emergency, chest pain, difficulty breathing, suicidal thoughts, self-harm), acknowledge calmly and immediately call escalate_to_human with urgency="high".

# SCHEDULING A CALL
- If the member asks to be called (e.g. "Can you call me?", "I'd prefer a phone call", "Please call me back"), acknowledge their request and call the schedule_call function using the phone number from their profile in your memory context.

# KNOWLEDGE BASE
- Use the search_knowledge_base function when the member asks about specific health policies, OTC card, clinical guidelines, medication information, eligibility, or any topic requiring accurate reference data.
- Do not guess at clinical or policy details — search the knowledge base first, then answer based on what you find.
- Keep your spoken response concise — summarize the key points from the search result rather than reading it verbatim.

# GOAL
Guide the conversation naturally toward a resolution while remaining concise enough for a high-quality audio experience."""


def _read_file(name: str) -> str:
    p = _DIR / name
    return p.read_text().strip() if p.exists() else ""


def get_inbound_system_prompt() -> str:
    inbound = _read_file("system_prompt_inbound.txt")
    if inbound:
        return f"{BASE_SYSTEM_PROMPT}\n\n{inbound}"
    return BASE_SYSTEM_PROMPT


def get_sms_system_prompt() -> str:
    sms = _read_file("system_prompt_sms.txt")
    if sms:
        return sms
    return (
        "You are an Owl Health care coordination agent continuing a text message conversation with a member. "
        "Be warm and natural — write like a text message. Short, plain sentences. No markdown."
    )


def get_outbound_system_prompt(name: str, goal: str, goal_desc: str) -> str:
    lines = [f"You are calling {name} on behalf of the Owl Health care team."]
    if goal:
        lines.append(f"The purpose of this call is: {goal}.")
    if goal_desc:
        lines.append(f"Follow-up guidance: {goal_desc}")
    call_prompt = "\n".join(lines)
    return f"{BASE_SYSTEM_PROMPT}\n\n{call_prompt}"
